### ISSUE TITLE
Validating IoU with recall

### ISSUE THREAD

**Issue by rafaelpadilla** [pjreddie/darknet#515]
I found that the **./darknet detector recall** function verifies the Intersection Over Union in a strange way.

It points out to the function validate_detector_recall in [Line 491 file detector.c](https://github.com/pjreddie/darknet/blob/master/examples/detector.c#L490) . 

It seems not to taking into account the class of the detected bounding box. It means that if an object of the class "dog" (for instance), is classified as a "house", but its bounding box intersects well with its ground truth, will result in a high IoU.

It also sets the threshold to 0.001 and the nms to 0.4, which are values that will lean to a high score.

Please, correct me if I'm wrong, but shouldn't the class of the detected object be considered?

Mr. Pad

**Comment by AlexeyAB** [comment]
@rafaelpadilla You are right. `./darknet detector recall` gives internal accuracy indicators that can not be compared with generally accepted indicators.

If you want to get mAP, IoU, Precision, Recall, F1 - calculated in a conventional way, you can use this fork: https://github.com/AlexeyAB/darknet

And use this commend: `./darknet detector map data/obj.data yolo-obj.cfg backup\yolo-obj_7000.weights`

**Comment by rafaelpadilla** [comment]
@AlexeyAB Thanks for your reply, but how does your fork calculate IoU? Does it take into consideration the class of the detected object? "cats" will be compared to "cats" and "dogs" will be compared to "dogs"? 

What happens when there are two or more bounding boxes over the same detection?

**Comment by AlexeyAB** [comment]
@rafaelpadilla 
> Does it take into consideration the class of the detected object?

Yes. 
If detected "cat" isn't overlaped with "cat" (or overlaped with "dog"), then this is false_positive and current_IoU = 0.

> What happens when there are two or more bounding boxes over the same detection?

Then only maximum IoU will be added to IoU_sum.
All non-maximum IoUs = 0 and they are false_positives.

And at the end:
`IoU_avg = IoU_sum / (True_positives + False_positives);`

----

Source code: https://github.com/AlexeyAB/darknet/blob/100d6f78011f0a773442411e2882a0203d390585/src/detector.c#L649-L704

And: https://github.com/AlexeyAB/darknet/blob/100d6f78011f0a773442411e2882a0203d390585/src/detector.c#L715

**Comment by sergYaNN** [comment]
@AlexeyAB  Hello Alexey, I am wondering, is there any way to perform estimation on validatoin set during the training. So that we can save the weights if we have improvement on the validation set.

**Comment by AlexeyAB** [comment]
@sergYaNN Isn't yet.

**Comment by sergYaNN** [comment]
@AlexeyAB  Thank you for the reply. I have impelemented smth on V2, and now working on that for V3. I am passing the network to a "validation" function and performing validation and deciding to save the weights or not. However, there is new function fuse_conv_batchnorm instead of set_batch_network(&net, 1).  Can you please explain  the purpose of merging batch_norm layers with conv layers for evaluation?

**Comment by AlexeyAB** [comment]
@sergYaNN 

>  However, there is new function fuse_conv_batchnorm instead of set_batch_network(&net, 1).

There is `network net = parse_network_cfg_custom(cfgfile, 1);`  instead of 2 lines:
```
network net = parse_network_cfg(cfgfile);
set_batch_network(&net, 1);
```

But `fuse_conv_batchnorm(net);` is for merging Convolutional and Batch-norm layers, it increases performance +7% for Detection. But it can't be applied for training.

If you want to call Validation from the Training, just remove or comment this function `fuse_conv_batchnorm(net);`

**Comment by sergYaNN** [comment]
@AlexeyAB  Okay, thanks.

**Comment by srhtyldz** [comment]
@AlexeyAB How did you draw the mAP chart while training ? i put ' -map ' while training but how could you get the chart ? can you please help for that ?

**Comment by adriacabeza** [comment]
@srhtyldz are you sure that you are using the @AlexeyAB's fork? Maybe you are still working on pjreddie's fork where it is not implemented.

**Comment by srhtyldz** [comment]
> @srhtyldz are you sure that you are using the @AlexeyAB's fork? Maybe you are still working on pjreddie's fork where it is not implemented.

Yes , I do. My question is about how can I get the chart ? where is it created in the Darknet folder ? I mean graph or chart.

**Comment by adriacabeza** [comment]
Hummmm @srhtyldz  I do not know it for sure because I have not checked all the source code, however, with a naked eye, I do not see any path where the chart is saved in the function ```float validate_detector_map(char *datacfg, char *cfgfile, char *weightfile, float thresh_calc_avg_iou, const float iou_thresh, const int map_points, network *existing_net)```. 

Please @AlexeyAB  correct me if I am wrong

**Comment by srhtyldz** [comment]
I got the chart like that 
but I don't know if its good or not .I couldn't understand that.Is that normal ?

![chart_1](https://user-images.githubusercontent.com/44969536/56643794-0446ab80-6683-11e9-81b7-fea81a3eb873.png)

**Comment by yudapp** [comment]
@srhtyldz did you find how to get the mAP chart? I did -map on a remote server but did not get anything. Is it saved somewhere, the mAP, Iou etc?

**Comment by priyankasinghvi** [comment]
@AlexeyAB I tried using detector map and detector recall. But I am unable to print graphs. It shows the error as follows :
It says that the Label file name is too short. Can't open label file name. I want to be able to plot Confusion Matrix, IoU, classification/misclassification etc. How can I do the same? 
Though I am able to plot mAP with the training loss.
![Capture00](https://user-images.githubusercontent.com/17821333/71093404-65d6ff80-21a9-11ea-8bb5-d60158323741.PNG)
Thank you

**Comment by feulhak** [comment]
did anyone know how to draw or plot the P-R curve for yolo prediction output? most papers show their P-R diagram but I do not know which function of the AlexeyABs repo they use?

**Comment by LucWuytens** [comment]
> Then only maximum IoU will be added to IoU_sum.
> All non-maximum IoUs = 0 and they are false_positives.
@AlexeyAB  
Hi Alexey, I have question about your statements above. You mention that all non-maximum IoUs = 0 and are false positives. Do you mean this only when those boxes predict the same class? Isn't non-max surpression supposed to prevent these alternative boxes of occurring before it is classified as false positive? If you mean two different boxes for two different classes, then I have an additional question: When the object is a cat and two boxes are predicted, one box says dog and the other one says cat. If the dog prediction has the highest class confidence score, so higher than the cat, then the dog prediction results in a false positive for the first box. However the second box is still considered as a True Positive in the output of 'darknet.exe detector map' according to my experiments. Can you confirm that? What I don't know for sure is how this second-box 'lower ranked' TP prediction influences mAP? I assume it is just ranked (as a TP) like other predictions and therefor part of mAP calculations. 
thanks for reply,
Luc
